src/h5web/vis-packs/core/heatmap/utils.ts

```diff
@@ -1,11 +1,44 @@
 import { range } from 'lodash-es';
 import type ndarray from 'ndarray';
+import {
+  BoundError,
+  CustomDomain,
+  Domain,
+  DomainErrors,
+  ScaleType,
+} from '../models';
 import { INTERPOLATORS } from './interpolators';
 import type { ColorMap, D3Interpolator, Dims } from './models';
 
 const GRADIENT_PRECISION = 1 / 20;
 const GRADIENT_RANGE = range(0, 1 + GRADIENT_PRECISION, GRADIENT_PRECISION);
 
+export function getVisDomain(
+  dataDomain: Domain,
+  customDomain: CustomDomain,
+  scaleType: ScaleType
+): [Domain, DomainErrors] {
+  const visDomain: Domain = [
+    customDomain[0] ?? dataDomain[0],
+    customDomain[1] ?? dataDomain[1],
+  ];
+
+  if (scaleType === ScaleType.Log) {
+    return [
+      [
+        visDomain[0] <= 0 ? dataDomain[0] : visDomain[0],
+        visDomain[1] <= 0 ? dataDomain[1] : visDomain[1],
+      ],
+      {
+        minError: visDomain[0] <= 0 ? BoundError.InvalidWithLog : undefined,
+        maxError: visDomain[1] <= 0 ? BoundError.InvalidWithLog : undefined,
+      },
+    ];
```

Comment:
Probably not the most clever code, but it will evolve once we start handling more errors.